// eslint-plugin-css/dist/utils/casing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.kebabCase = kebabCase;
exports.isKebabCase = isKebabCase;
exports.camelCase = camelCase;
exports.isCamelCase = isCamelCase;
exports.getChecker = getChecker;
exports.getConverter = getConverter;
exports.getExactConverter = getExactConverter;
function hasSymbols(str) {
    return /[!"#%&'()*+,./:;<=>?@[\\\]^`{|}]/u.exec(str);
}
function hasUpper(str) {
    return /[A-Z]/u.exec(str);
}
function kebabCase(str) {
    return str
        .replace(/\B(?<c>[A-Z])/gu, "-$<c>")
        .toLowerCase()
        .replace(/^(?<vender>apple|moz|ms|o|wap|webkit)-/u, "-$<vender>-");
}
function isKebabCase(str) {
    if (hasUpper(str) || hasSymbols(str) || /_|--|\s/u.test(str)) {
        return false;
    }
    return true;
}
function camelCase(str) {
    return str
        .replace(/-(?<c>\w)/gu, (_, c) => (c ? c.toUpperCase() : ""))
        .replace(/^Ms[A-Z]/u, (ms) => `m${ms.slice(1)}`);
}
function isCamelCase(str) {
    if (hasSymbols(str) ||
        /[\s\-_]/u.test(str)) {
        return false;
    }
    return true;
}
const convertersMap = {
    "kebab-case": kebabCase,
    camelCase,
};
const checkersMap = {
    "kebab-case": isKebabCase,
    camelCase: isCamelCase,
};
function getChecker(name) {
    return checkersMap[name] || isCamelCase;
}
function getConverter(name) {
    return convertersMap[name] || camelCase;
}
function getExactConverter(name) {
    const converter = getConverter(name);
    const checker = getChecker(name);
    return (str) => {
        const result = converter(str);
        return checker(result) ? result : str;
    };
}
