// eslint-plugin-css/dist/utils/ast-utils/utils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.getParent = getParent;
exports.findVariable = findVariable;
exports.getPropertyName = getPropertyName;
exports.getStaticValue = getStaticValue;
exports.getScope = getScope;
exports.findExpression = findExpression;
exports.isStaticTemplateLiteral = isStaticTemplateLiteral;
exports.isStringLiteral = isStringLiteral;
const eslintUtils = __importStar(require("@eslint-community/eslint-utils"));
function getParent(node) {
    if (!node) {
        return null;
    }
    return node.parent;
}
function findVariable(context, node) {
    return eslintUtils.findVariable(getScope(context, node), node);
}
function getPropertyName(context, node) {
    var _a;
    return (_a = eslintUtils.getPropertyName(node, getScope(context, node))) !== null && _a !== void 0 ? _a : null;
}
function getStaticValue(context, node) {
    return eslintUtils.getStaticValue(node, getScope(context, node));
}
function getScope(context, currentNode) {
    var _a;
    const sourceCode = (_a = context.sourceCode) !== null && _a !== void 0 ? _a : context.getSourceCode();
    const scopeManager = sourceCode.scopeManager;
    let node = currentNode;
    for (; node; node = node.parent || null) {
        const scope = scopeManager.acquire(node, false);
        if (scope) {
            if (scope.type === "function-expression-name") {
                return scope.childScopes[0];
            }
            return scope;
        }
    }
    return scopeManager.scopes[0];
}
function findExpression(context, id) {
    let target = id;
    const set = new Set();
    while (target.type === "Identifier") {
        if (set.has(target)) {
            return null;
        }
        set.add(target);
        const variable = findVariable(context, target);
        if (!variable) {
            return null;
        }
        if (variable.defs.length !== 1) {
            return null;
        }
        const def = variable.defs[0];
        if (def.type !== "Variable" ||
            def.parent.kind !== "const" ||
            !def.node.init) {
            return null;
        }
        target = def.node.init;
    }
    return target;
}
function isStaticTemplateLiteral(node) {
    return node.type === "TemplateLiteral" && node.quasis.length === 1;
}
function isStringLiteral(node) {
    return node.type === "Literal" && typeof node.value === "string";
}
